File: ios/Cloudmoji/Cloudmoji/Views/Launcher/LauncherTile.swift
import SwiftUI
import CloudmojiCore

/// Every number a launcher tile is drawn from.
///
/// Recorded in `docs/design/DESIGN_SYSTEM.md` alongside the emoji tile and the
/// count tile, for the same reason those are: the touch-target rule in
/// `CLAUDE.md` is a rule rather than a literal, so it has to be assertable.
enum LauncherTileMetrics {
    /// The visible squircle is Home-Screen sized, while the whole icon-and-label
    /// cell remains the much larger child-facing target.
    static let iconSide: CGFloat = 76
    static let compactIconSide: CGFloat = 68
    static let cellHeight: CGFloat = 112
    static let compactCellHeight: CGFloat = 94

    /// `CLAUDE.md` rule 2, the floor between adjacent child-facing targets.
    /// Doubles as the grid's row and column spacing.
    static let spacing: CGFloat = 10

    static let cornerRadius: CGFloat = 22

    static let labelSize: CGFloat = 13
    static let compactLabelSize: CGFloat = 12

    /// Design system Active States: tabs `scale(0.9)`. A launcher tile is the
    /// tab bar's successor, so it takes the tab's number rather than inventing
    /// an fifth one.
    static let pressedScale: CGFloat = 0.92

}

/// One mini-app on the launcher.
///
/// Its own type, like `EmojiTile` and `CountTile`, so the child-facing 64pt
/// floor can be measured off a real render rather than read back out of the
/// constant that set it.
struct LauncherTile: View {
    let app: MiniApp
    /// Already resolved to the family's language by the caller — the tile does
    /// not read settings, the same way no other tile in this app does.
    let label: String
    var isCompact: Bool = false
    let onTap: () -> Void

    private var iconSide: CGFloat {
        isCompact ? LauncherTileMetrics.compactIconSide : LauncherTileMetrics.iconSide
    }

    private var cellHeight: CGFloat {
        isCompact ? LauncherTileMetrics.compactCellHeight : LauncherTileMetrics.cellHeight
    }

    var body: some View {
        Button {
            // Before anything else, so the buzz lands with the finger rather
            // than after the incoming screen has decided what to draw.
            Haptics.tap()
            onTap()
        } label: {
            VStack(spacing: isCompact ? 4 : 7) {
                LauncherAppIcon(app: app, side: iconSide)

                Text(label)
                    .font(Theme.body(isCompact
                                     ? LauncherTileMetrics.compactLabelSize
                                     : LauncherTileMetrics.labelSize, .black))
                    // Text, so it takes the button's accent tint — the system
                    // blue — unless this says otherwise. The glyph above is a
                    // colour emoji and is unaffected.
                    .foregroundStyle(Theme.textPrimary)
                    .lineLimit(1)
                    .minimumScaleFactor(0.7)
                    .shadow(color: Theme.bgPrimary.opacity(0.9), radius: 2, y: 1)
            }
            .frame(maxWidth: .infinity, minHeight: cellHeight)
            .contentShape(Rectangle())
        }
        .buttonStyle(PressScale(scale: LauncherTileMetrics.pressedScale))
        .accessibilityLabel(label)
        .accessibilityIdentifier("launcher-tile-\(app.rawValue)")
    }
}

/// A layered, in-app app icon.
///
/// It follows the visual grammar of an iOS Home Screen icon — one simple idea,
/// a dimensional background and a consistent squircle — without pretending to
/// be a system-owned icon or embedding tiny interface screenshots.
struct LauncherAppIcon: View {
    let app: MiniApp
    let side: CGFloat

    private var shape: RoundedRectangle {
        RoundedRectangle(
            cornerRadius: side * 0.25,
            style: .continuous
        )
    }

    var body: some View {
        ZStack {
            shape
                .fill(
                    LinearGradient(
                        colors: [
                            app.visualTheme.accent,
                            app.visualTheme.secondary,
                            Theme.bgMid.opacity(0.88),
                        ],
                        startPoint: .topLeading,
                        endPoint: .bottomTrailing
                    )
                )

            Circle()
                .fill(Color.white.opacity(0.18))
                .frame(width: side * 0.76)
                .blur(radius: side * 0.12)
                .offset(x: -side * 0.25, y: -side * 0.28)

            artwork
                .frame(width: side, height: side)

            shape
                .strokeBorder(Color.white.opacity(0.24), lineWidth: 1)
        }
        .frame(width: side, height: side)
        .clipShape(shape)
        .shadow(color: app.visualTheme.accent.opacity(0.28), radius: 10, y: 6)
        .accessibilityHidden(true)
    }

    @ViewBuilder private var artwork: some View {
        switch app {
        case .words:
            ZStack {
                Image(systemName: "quote.bubble.fill")
                    .font(.system(size: side * 0.43, weight: .black))
                    .foregroundStyle(Color.white)
                Circle()
                    .fill(Theme.coral)
                    .frame(width: side * 0.18)
                    .overlay(
                        Image(systemName: "waveform")
                            .font(.system(size: side * 0.09, weight: .black))
                            .foregroundStyle(Color.white)
                    )
                    .offset(x: side * 0.24, y: side * 0.23)
            }

        case .count:
            ZStack {
                RoundedRectangle(cornerRadius: side * 0.16, style: .continuous)
                    .fill(Color.white.opacity(0.92))
                    .frame(width: side * 0.58, height: side * 0.58)
                    .rotationEffect(.degrees(-7))
                Text("123")
                    .font(Theme.body(side * 0.24, .black))
                    .foregroundStyle(Theme.bgMid)
            }

        case .flashCards:
            ZStack {
                card(color: Color.white.opacity(0.34))
                    .rotationEffect(.degrees(-13))
                    .offset(x: -side * 0.08, y: side * 0.02)
                card(color: Color.white.opacity(0.94))
                    .rotationEffect(.degrees(7))
                    .offset(x: side * 0.06)
                Image(systemName: "bolt.fill")
                    .font(.system(size: side * 0.34, weight: .black))
                    .foregroundStyle(Theme.coral)
                    .rotationEffect(.degrees(7))
            }

        case .instrument:
            ZStack {
                HStack(spacing: side * 0.035) {
                    ForEach(0..<4, id: \.self) { index in
                        RoundedRectangle(cornerRadius: side * 0.035)
                            .fill(index.isMultiple(of: 2) ? Color.white : Theme.gold)
                    }
                }
                .frame(width: side * 0.64, height: side * 0.44)
                .offset(y: side * 0.08)
                Image(systemName: "music.note")
                    .font(.system(size: side * 0.28, weight: .black))
                    .foregroundStyle(Color.white)
                    .offset(x: side * 0.18, y: -side * 0.22)
            }

        case .animalSounds:
            ZStack {
                Image(systemName: "pawprint.fill")
                    .font(.system(size: side * 0.46, weight: .black))
                    .foregroundStyle(Color.white)
                Image(systemName: "waveform")
                    .font(.system(size: side * 0.19, weight: .black))
                    .foregroundStyle(Theme.gold)
                    .offset(x: side * 0.25, y: -side * 0.24)
            }

        case .photos:
            ZStack {
                Image(systemName: "camera.fill")
                    .font(.system(size: side * 0.45, weight: .black))
                    .foregroundStyle(Color.white)
                Circle()
                    .fill(Theme.bgMid)
                    .frame(width: side * 0.21)
                Circle()
                    .stroke(Theme.moonlight, lineWidth: side * 0.035)
                    .frame(width: side * 0.12)
            }

        case .sleepy:
            ZStack {
                Image(systemName: "moon.fill")
                    .font(.system(size: side * 0.47, weight: .black))
                    .foregroundStyle(Color.white)
                    .offset(x: -side * 0.08, y: -side * 0.08)
                Image(systemName: "cloud.fill")
                    .font(.system(size: side * 0.31, weight: .black))
                    .foregroundStyle(Theme.moonlight)
                    .offset(x: side * 0.16, y: side * 0.20)
            }
        }
    }

    private func card(color: Color) -> some View {
        RoundedRectangle(cornerRadius: side * 0.09, style: .continuous)
            .fill(color)
            .frame(width: side * 0.48, height: side * 0.58)
    }
}

#Preview("Home Screen icons") {
    ZStack {
        Theme.background.ignoresSafeArea()
        LazyVGrid(
            columns: Array(repeating: GridItem(.flexible(), spacing: 10), count: 4),
            spacing: 10
        ) {
            ForEach(MiniApp.allCases) { app in
                LauncherTile(
                    app: app,
                    label: app.label(.en),
                    isCompact: false
                ) {}
            }
        }
        .padding()
    }
}

File: android/app/src/main/java/app/cloudmoji/android/ui/launcher/LauncherTile.kt
package app.cloudmoji.android.ui.launcher

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.cloudmoji.android.model.MiniApp
import app.cloudmoji.android.ui.common.pressScale
import app.cloudmoji.android.ui.theme.Amber
import app.cloudmoji.android.ui.theme.BackgroundMid
import app.cloudmoji.android.ui.theme.CloudmojiBodyFont
import app.cloudmoji.android.ui.theme.Coral
import app.cloudmoji.android.ui.theme.Gold
import app.cloudmoji.android.ui.theme.Lavender
import app.cloudmoji.android.ui.theme.Moonlight
import app.cloudmoji.android.ui.theme.Teal
import app.cloudmoji.android.ui.theme.TextPrimary

/**
 * Every number a launcher tile is drawn from. Mirrors iOS
 * `LauncherTile.swift`'s `LauncherTileMetrics`, dp for pt.
 *
 * Sized as step functions of `(compact, isExpandedPad)` — the same shape as
 * `app.cloudmoji.android.ui.count.CountTileMetrics` — rather than a single
 * constant, so [LauncherTileMetricsTest] can assert the child-target floor
 * across every combination the real [app.cloudmoji.android.ui.common.CloudmojiLayout]
 * can ever hand this screen, not just the one shape a device happens to be
 * running.
 */
object LauncherTileMetrics {
    /** `CLAUDE.md` rule 1's floor. The *cell* — icon plus caption, the whole
     * tappable region [LauncherTile] draws — is what has to clear this, not
     * the visible squircle alone: iOS's own comment on `LauncherTile.swift`
     * says the same thing ("the visible squircle is Home-Screen sized, while
     * the whole icon-and-label cell remains the much larger child-facing
     * target"). */
    val childMinimum: Dp = 64.dp

    /** `CLAUDE.md` rule 2's floor between adjacent child-facing targets. */
    val gapMinimum: Dp = 8.dp

    fun iconSide(compact: Boolean, isExpandedPad: Boolean): Dp = when {
        isExpandedPad -> 118.dp
        compact -> 68.dp
        else -> 76.dp
    }

    fun cellHeight(compact: Boolean, isExpandedPad: Boolean): Dp = when {
        isExpandedPad -> 164.dp
        compact -> 94.dp
        else -> 112.dp
    }

    fun spacing(isExpandedPad: Boolean): Dp = if (isExpandedPad) 20.dp else 10.dp

    fun labelSize(compact: Boolean, isExpandedPad: Boolean): TextUnit = when {
        isExpandedPad -> 17.sp
        compact -> 12.sp
        else -> 13.sp
    }

    /** Caps the grid's own width and centers it, standing in for iOS's fixed
     * iPad cell width (140pt) — both exist for the same reason: stop icons
     * spreading edge-to-edge on a wide screen. */
    fun maxGridWidth(isExpandedPad: Boolean): Dp = if (isExpandedPad) 780.dp else 460.dp

    /** `docs/design/DESIGN_SYSTEM.md`'s "22px — 76pt launcher icon (iOS);
     * computed as 25% of the icon side" — the fraction is the actual rule
     * iOS's `LauncherAppIcon.shape` applies (`side * 0.25`); its sibling
     * `cornerRadius: CGFloat = 22` constant is dead code on the iOS side
     * (never referenced by any shape) and is deliberately not ported. */
    const val cornerRadiusFraction: Float = 0.25f

    /** Design system Active States: "Launcher icon cells: `scale(0.92)`". */
    const val pressedScale: Float = 0.92f
}

/** The visual identity of one mini-app's layered launcher icon — accent and
 * secondary stops of the icon's gradient. Mirrors iOS `MiniAppVisualTheme`,
 * narrowed to the two colors [LauncherAppIcon] actually draws with (the third
 * stop is always the app's own background, not a per-app color). Kept public
 * (like `ui.mascot.MascotStyle`) so [LauncherTileMetricsTest] can assert the
 * mapping directly rather than only through a rendered tree. */
data class LauncherIconTheme(val accent: Color, val secondary: Color)

/** Ported from iOS `MiniApp.visualTheme`, case for case. Photos' secondary
 * stop is [Moonlight] there — the Phase 0 Android tile had drifted to
 * [Amber], which this corrects. */
fun MiniApp.launcherIconTheme(): LauncherIconTheme = when (this) {
    MiniApp.Words -> LauncherIconTheme(Teal, Coral)
    MiniApp.Count -> LauncherIconTheme(Gold, Amber)
    MiniApp.FlashCards -> LauncherIconTheme(Gold, Lavender)
    MiniApp.Music -> LauncherIconTheme(Coral, Moonlight)
    MiniApp.Animals -> LauncherIconTheme(Teal, Gold)
    MiniApp.Photos -> LauncherIconTheme(Coral, Moonlight)
    MiniApp.Sleepy -> LauncherIconTheme(Moonlight, Lavender)
}

/**
 * One mini-app on the launcher: a layered squircle icon above a one-line
 * caption. Ported from iOS `LauncherTile.swift`.
 *
 * [label] arrives already resolved to the family's language — this tile does
 * not read settings, the same way no other tile in the app does.
 */
@Composable
fun LauncherTile(
    app: MiniApp,
    label: String,
    iconSide: Dp,
    cellHeight: Dp,
    labelSize: TextUnit,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val interactionSource = remember { MutableInteractionSource() }

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
        modifier = modifier
            .fillMaxWidth()
            .height(cellHeight)
            .pressScale(interactionSource, LauncherTileMetrics.pressedScale)
            .clickable(
                interactionSource = interactionSource,
                indication = null,
                role = Role.Button,
                onClick = onClick,
            )
            .semantics {
                role = Role.Button
                contentDescription = label
            }
            .testTag("launcher-tile-${app.route}"),
    ) {
        LauncherAppIcon(app = app, side = iconSide)

        Spacer(Modifier.height(7.dp))
        Text(
            text = label,
            color = TextPrimary,
            fontFamily = CloudmojiBodyFont,
            fontWeight = FontWeight.Black,
            fontSize = labelSize,
            maxLines = 1,
            textAlign = TextAlign.Center,
        )
    }
}

/**
 * A layered, in-app app icon: an accent-to-secondary gradient squircle, a
 * soft highlight in the top-left corner standing in for iOS's blurred white
 * circle (`Modifier.blur` needs `RenderEffect`, API 31+; this module's
 * `minSdk` is 26 — the same trade-off `ui/mascot/CloudMascot.kt` already
 * documents for its own ambient shadow), the mini-app's emoji glyph, and a
 * hairline border. Ported from iOS `LauncherAppIcon`.
 */
@Composable
private fun LauncherAppIcon(app: MiniApp, side: Dp) {
    val theme = remember(app) { app.launcherIconTheme() }
    val shape = remember(side) { RoundedCornerShape(side * LauncherTileMetrics.cornerRadiusFraction) }
    val density = LocalDensity.current
    val sidePx = with(density) { side.toPx() }
    val highlight = remember(sidePx) {
        Brush.radialGradient(
            colors = listOf(Color.White.copy(alpha = 0.22f), Color.White.copy(alpha = 0f)),
            center = Offset(sidePx * 0.26f, sidePx * 0.24f),
            radius = sidePx * 0.55f,
        )
    }

    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(side)
            .shadow(10.dp, shape)
            .clip(shape)
            .background(Brush.linearGradient(listOf(theme.accent, theme.secondary, BackgroundMid.copy(alpha = 0.88f))))
            .background(highlight)
            .border(1.dp, Color.White.copy(alpha = 0.24f), shape),
    ) {
        Text(text = app.icon, fontSize = (side.value * 0.46f).sp, textAlign = TextAlign.Center)
    }
}
